server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java

```diff
@@ -408,6 +409,10 @@ static int shortcutTotalHitCount(IndexReader reader, Query query) throws IOExcep
                 FieldInfos fieldInfos = context.reader().getFieldInfos();
                 FieldInfo fieldInfo = fieldInfos.fieldInfo(field);
                 if (fieldInfo != null) {
+                    if (fieldInfo.getDocValuesType() == DocValuesType.NONE) {
+                        // no shortcut possible: it's a text field, empty values are counted as no value.
```

Comment:
For my understanding of the comment: do we know that we have a text field here because its the only one without doc values or do we know this from somewhere else? Or asked another way: could there be other field types without doc values that are not text, in which case the comment can probably be more general

Reply:
Good question, this is not straight-forward: there are certainly other field types that may not have doc_values, e.g. a keyword with doc_values disabled, although not so common. At the lucene level we don't really have a way to distinguish between the two. What looked funny here is that we check the doc_values type, and then if there are doc_values we may look at terms, that have nothing to do with doc_values. This is to ensure that we don't shortcut for text fields, as empty tokens after analysis means no value, like the comment says.

For more info, this logic emulates what the [`FieldExistsQuery`](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/FieldExistsQuery.java#L229) in Lucene does.